### PR TITLE
+h2 handle MAX_FRAME_SIZE, merge framing stages into bidi

### PR DESCRIPTION
**Pull request by ktoso**
- Solves part of https://github.com/akka/akka-http/issues/757 - the DATA part of it, we buffer data until the entire thing was pushed out (chunked up into at-most-MAX_FRAME_SIZE).
- GOAWAYs wrong set frame sizes

### CODE REVIEW

**Comment by ktoso** [comment]
Now also handling https://github.com/akka/akka-http/issues/776 SETTINGS_HEADER_TABLE_SIZE in hpack
Over there it would also be nice to be able to swallow it early but that would need a bidistage.

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/FrameEvent.scala [hunk 1]
```diff
@@ -71,3 +73,10 @@ final case class ParsedHeadersFrame(
   keyValuePairs: Seq[(String, String)],
   priorityInfo:  Option[PriorityFrame]
 ) extends FrameEvent
+
+// ------ internal frames, used for communicating settings between stages of our pipeline ----- 
+/**
+ * Special synthetic FrameEvent used to communicate enoder setting changes back to encoder.
+ * MUST NOT be rendered to network.
+ */
+private[http] final case class SyntheticHpackEncoderSettingFrame(setting: Setting) extends SyntheticFrameEvent
```
Comment (ktoso) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/FrameEvent.scala]: Either we do this dance with sending back "down" the setting in a special one like that, or we need to make the HPack stuff a bidistage the same way like the framing was done now.

**Comment by ktoso** [comment]
Best to review the 2 commits at the same time, not independently IMO. Notice that in hpack where I did not make it a bidistage we need to do the weird "bounce the setting back down for the encoder to apply it" which is basically what we've avoided in framing by making it a bidistage. 

I don't have a good test for setting the table size yet hm...

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2Protocol.scala [hunk 2]
```diff
@@ -191,7 +192,9 @@ object Http2Protocol {
     def byId(id: Int): SettingIdentifier = All(id - 1)
   }
 
-  sealed abstract class ErrorCode(val id: Int) extends Product
+  sealed abstract class ErrorCode(val id: Int) extends Product {
+    def name = Logging.simpleName(this).replaceAll("$", "")
```
Comment (jrudolph) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2Protocol.scala]: Why not `productPrefix`?

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2ServerDemux.scala [hunk 3]
```diff
@@ -114,11 +114,11 @@ class Http2ServerDemux extends GraphStage[BidiShape[Http2SubStream, FrameEvent,
         incomingStreams.keys.toList.sortBy(-_).headOption.getOrElse(0) // FIXME should be optimised
       }
 
-      def pushGOAWAY(errorCode: ErrorCode = Http2Protocol.ErrorCode.PROTOCOL_ERROR): Unit = {
+      def pushGOAWAY(errorCode: ErrorCode = Http2Protocol.ErrorCode.PROTOCOL_ERROR, debug: String = ""): Unit = {
```
Comment (jrudolph) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2ServerDemux.scala]: Good one, we should annotate all occurrences with this debug info, not only for the peer but also to log it on our side.

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2ServerDemux.scala [hunk 4]
```diff
@@ -179,32 +179,51 @@ class Http2ServerDemux extends GraphStage[BidiShape[Http2SubStream, FrameEvent,
               incomingStreams(streamId).inlet.foreach(_.cancel())
 
             case WindowUpdateFrame(streamId, increment) ⇒
-              incomingStreams(streamId).outboundWindowLeft += increment
+              incomingStreams(streamId).outboundWindowLeft = incomingStreams(streamId).outboundWindowLeft + increment
```
Comment (jrudolph) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2ServerDemux.scala]: Does that change anything? (Other than doing the map lookup twice...)

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/hpack/HeaderCompression.scala [hunk 5]
```diff
@@ -5,39 +5,45 @@
...
 
     object Idle extends State {
-      val handleEvent: PartialFunction[FrameEvent, Unit] = {
+      override val handleEvent: PartialFunction[FrameEvent, Unit] = {
         case fr @ ParsedHeadersFrame(streamId, endStream, kvs, prioInfo) ⇒
           os.reset()
           kvs.foreach {
             case (key, value) ⇒
-              encoder.encodeHeader(os, key.getBytes(HeaderDecompression.UTF8), value.getBytes(HeaderDecompression.UTF8), false)
+              encoder.encodeHeader(os, key.getBytes(StandardCharsets.UTF_8), value.getBytes(StandardCharsets.UTF_8), false)
```
Comment (jrudolph) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/hpack/HeaderCompression.scala]: Nice, thanks. Forgot about these.

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala [hunk 6]
```diff
@@ -500,23 +502,113 @@ class Http2ServerSpec extends AkkaSpec with WithInPendingUntilFixed with Eventua
         error should ===(ErrorCode.FRAME_SIZE_ERROR)
       }
 
-      "received SETTINGS_MAX_FRAME_SIZE" in pending
+      "received SETTINGS_MAX_FRAME_SIZE should cause outgoing DATA to be chunked up into at-most-that-size parts " in new TestSetup with RequestResponseProbes {
```
Comment (jrudolph) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala]: nice test

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala [hunk 7]
```diff
@@ -500,23 +502,113 @@ class Http2ServerSpec extends AkkaSpec with WithInPendingUntilFixed with Eventua
...
+        def setMaxFrameSize(size: Int) = SettingsFrame(Setting(SettingIdentifier.SETTINGS_MAX_FRAME_SIZE, size) :: Nil)
```
Comment (jrudolph) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala]: Maybe add `sendSettingsFrame(settingIdentifier, value)` to the TestSetup, instead. Will be needed more often than that.

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/FrameEvent.scala (hunk 1 above)
Comment (jrudolph) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/FrameEvent.scala]: What if we change the `SettingsAckFrame` to carry all ack'd settings, instead? (We don't change the actual rendering, of course.)

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2ServerDemux.scala (hunk 4 above)
Comment (ktoso) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2ServerDemux.scala]: good point will revert back, I was debugging things here

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2Protocol.scala (hunk 2 above)
Comment (ktoso) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2Protocol.scala]: Good one 👍

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/FrameEvent.scala (hunk 1 above)
Comment (ktoso) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/FrameEvent.scala]: Yeah could be nicer indeed 👍

**Comment by jrudolph** [comment]
Regarding bouncing settings vs. handling it right where it is applied: I think there might be places where it could be worthwhile if all stages see a settings change even if not directly affected (e.g. demux would probably want to try to buffer full frames for outgoing data but not more).

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala (hunk 7 above)
Comment (ktoso) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala]: Yup, sure 👍

**Comment by ktoso** [comment]
// Talked it over a bit on chat.
Yeah let's go back to the bouncing back style. It will work out simpler than initially expected.
I'll change the Ack to carry around the setting, such that we can apply it on the "way out".
Seems we'll still need the synthetic version though, but I'll give it some more thought.

The spec does not really require "immediate" applying however it does in-order, which we can keep using the bouncing. Passing around the acks could impact perf a bit but it'll in no way be a big part of perf so I was over worrying perhaps.

Moving to bouncing style then.

Also added some more fun tests, adding now that we must now overflow the demand window etc.

**Comment by jrudolph** [comment]
I'm closing this one as I think should break the parts of it apart to resolve the conflicts and actually make it mergeable.

**Comment by jrudolph** [comment]
I extracted #786 and #787 from this PR. Unfortunately, these only contained the small and / or cosmetic changes.

All the remaining changes are lumped together in this commit: https://github.com/jrudolph/akka-http/commit/a3214dfb821f7f6498b4d7b6bf0a1a0323f6cb22, let's start again from that state.

**Comment by ktoso** [comment]
Ok, thanks for breaking it up like that!
